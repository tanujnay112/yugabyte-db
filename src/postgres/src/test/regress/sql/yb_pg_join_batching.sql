CREATE TABLE p1 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p1 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 2 = 0;
ANALYZE p1;

CREATE TABLE p2 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p2 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 3 = 0;
ANALYZE p2;

CREATE TABLE p3 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p3 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 5 = 0;

CREATE TABLE p4 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p4 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 7 = 0;
ANALYZE p4;

-- We're testing nested loop join batching in this file
SET enable_hashjoin = off;
SET enable_mergejoin = off;

SET yb_nl_batch_size = 3;

EXPLAIN (COSTS OFF) SELECT * FROM p1 t1 JOIN p2 t2 ON t1.a = t2.a WHERE t1.a <= 100 AND t2.a <= 100;
SELECT * FROM p1 t1 JOIN p2 t2 ON t1.a = t2.a WHERE t1.a <= 100 AND t2.a <= 100;

EXPLAIN (COSTS OFF) SELECT * FROM p3 t3 LEFT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.a <= 100 AND t2.a <= 100) s ON t3.a = s.a WHERE t3.a <= 30;
SELECT * FROM p3 t3 LEFT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.a <= 100 AND t2.a <= 100) s ON t3.a = s.a WHERE t3.a <= 30;

EXPLAIN (COSTS OFF) SELECT * FROM p3 t3 RIGHT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.b <= 10 AND t2.b <= 15) s ON t3.a = s.a;
SELECT * FROM p3 t3 RIGHT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.b <= 10 AND t2.b <= 15) s ON t3.a = s.a;

-- anti join--
EXPLAIN (COSTS OFF) SELECT * FROM p1 t1 WHERE NOT EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;
SELECT * FROM p1 t1 WHERE NOT EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;

-- semi join--
EXPLAIN (COSTS OFF) SELECT * FROM p1 t1 WHERE EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;
SELECT * FROM p1 t1 WHERE EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;

set yb_nl_batch_size to 10;
set enable_seqscan to false;
explain (costs off) select * from p1 a join p2 b on a.a = b.a join p3 c on b.a = c.a join p4 d on a.b = d.b where a.b = 10 ORDER BY a.a, b.a, c.a, d.a;
select * from p1 a join p2 b on a.a = b.a join p3 c on b.a = c.a join p4 d on a.b = d.b where a.b = 10 ORDER BY a.a, b.a, c.a, d.a;

DROP TABLE p1;
DROP TABLE p2;
DROP TABLE p3;
DROP TABLE p4;
