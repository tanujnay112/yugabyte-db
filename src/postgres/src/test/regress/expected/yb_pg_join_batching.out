CREATE TABLE p1 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p1 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 2 = 0;
ANALYZE p1;
CREATE TABLE p2 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p2 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 3 = 0;
ANALYZE p2;
CREATE TABLE p3 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p3 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 5 = 0;
ANALYZE p3;
CREATE TABLE p4 (a int, b int, c varchar, primary key(a,b));
INSERT INTO p4 SELECT i, i % 25, to_char(i, 'FM0000') FROM generate_series(0, 599) i WHERE i % 7 = 0;
ANALYZE p4;
-- We're testing nested loop join batching in this file
SET enable_hashjoin = off;
SET enable_mergejoin = off;
SET yb_nl_batch_size = 3;
EXPLAIN (COSTS OFF) SELECT * FROM p1 t1 JOIN p2 t2 ON t1.a = t2.a WHERE t1.a <= 100 AND t2.a <= 100;
                              QUERY PLAN                              
----------------------------------------------------------------------
 Nested Loop
   Join Filter: (t1.a = t2.a)
   ->  Seq Scan on p2 t2
         Filter: (a <= 100)
   ->  Index Scan using p1_pkey on p1 t1
         Index Cond: ((a = ANY (ARRAY[t2.a, $1, $2])) AND (a <= 100))
(6 rows)

SELECT * FROM p1 t1 JOIN p2 t2 ON t1.a = t2.a WHERE t1.a <= 100 AND t2.a <= 100;
 a  | b  |  c   | a  | b  |  c   
----+----+------+----+----+------
 78 |  3 | 0078 | 78 |  3 | 0078
 12 | 12 | 0012 | 12 | 12 | 0012
 90 | 15 | 0090 | 90 | 15 | 0090
  6 |  6 | 0006 |  6 |  6 | 0006
 42 | 17 | 0042 | 42 | 17 | 0042
 48 | 23 | 0048 | 48 | 23 | 0048
 96 | 21 | 0096 | 96 | 21 | 0096
 60 | 10 | 0060 | 60 | 10 | 0060
 72 | 22 | 0072 | 72 | 22 | 0072
 36 | 11 | 0036 | 36 | 11 | 0036
 54 |  4 | 0054 | 54 |  4 | 0054
 18 | 18 | 0018 | 18 | 18 | 0018
 66 | 16 | 0066 | 66 | 16 | 0066
 30 |  5 | 0030 | 30 |  5 | 0030
 84 |  9 | 0084 | 84 |  9 | 0084
  0 |  0 | 0000 |  0 |  0 | 0000
 24 | 24 | 0024 | 24 | 24 | 0024
(17 rows)

EXPLAIN (COSTS OFF) SELECT * FROM p3 t3 LEFT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.a <= 100 AND t2.a <= 100) s ON t3.a = s.a WHERE t3.a <= 30;
                                    QUERY PLAN                                    
----------------------------------------------------------------------------------
 Nested Loop Left Join
   Join Filter: (t3.a = t1.a)
   ->  Seq Scan on p3 t3
         Filter: (a <= 30)
   ->  Materialize
         ->  Nested Loop
               Join Filter: (t1.a = t2.b)
               ->  Seq Scan on p2 t2
                     Filter: (a <= 100)
               ->  Index Scan using p1_pkey on p1 t1
                     Index Cond: ((a = ANY (ARRAY[t2.b, $1, $2])) AND (a <= 100))
(11 rows)

SELECT * FROM p3 t3 LEFT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.a <= 100 AND t2.a <= 100) s ON t3.a = s.a WHERE t3.a <= 30;
 a  | b  |  c   | a  
----+----+------+----
  5 |  5 | 0005 |   
 15 | 15 | 0015 |   
 10 | 10 | 0010 | 10
 30 |  5 | 0030 |   
  0 |  0 | 0000 |  0
  0 |  0 | 0000 |  0
 25 |  0 | 0025 |   
 20 | 20 | 0020 | 20
(8 rows)

EXPLAIN (COSTS OFF) SELECT * FROM p3 t3 RIGHT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.b <= 10 AND t2.b <= 15) s ON t3.a = s.a;
                                QUERY PLAN                                 
---------------------------------------------------------------------------
 Nested Loop Left Join
   Join Filter: (t3.a = t1.a)
   ->  Nested Loop
         Join Filter: (t1.a = t2.b)
         ->  Seq Scan on p2 t2
               Filter: (b <= 15)
         ->  Index Scan using p1_pkey on p1 t1
               Index Cond: ((a = ANY (ARRAY[t2.b, $1, $2])) AND (b <= 10))
   ->  Index Scan using p3_pkey on p3 t3
         Index Cond: (a = ANY (ARRAY[t1.a, $4, $5]))
(10 rows)

SELECT * FROM p3 t3 RIGHT OUTER JOIN (SELECT t1.a as a FROM p1 t1 JOIN p2 t2 ON t1.a = t2.b WHERE t1.b <= 10 AND t2.b <= 15) s ON t3.a = s.a;
 a  | b  |  c   | a  
----+----+------+----
 10 | 10 | 0010 | 10
    |    |      |  6
    |    |      |  6
 10 | 10 | 0010 | 10
 10 | 10 | 0010 | 10
    |    |      |  8
    |    |      |  2
    |    |      |  2
    |    |      |  4
  0 |  0 | 0000 |  0
  0 |  0 | 0000 |  0
 10 | 10 | 0010 | 10
  0 |  0 | 0000 |  0
    |    |      |  8
    |    |      |  4
    |    |      |  6
    |    |      |  4
    |    |      |  4
 10 | 10 | 0010 | 10
 10 | 10 | 0010 | 10
    |    |      |  4
    |    |      |  8
    |    |      |  6
    |    |      |  2
    |    |      |  8
    |    |      |  2
    |    |      |  4
  0 |  0 | 0000 |  0
 10 | 10 | 0010 | 10
    |    |      |  6
  0 |  0 | 0000 |  0
    |    |      |  8
    |    |      |  4
  0 |  0 | 0000 |  0
  0 |  0 | 0000 |  0
    |    |      |  2
    |    |      |  4
    |    |      |  2
    |    |      |  8
    |    |      |  8
    |    |      |  2
    |    |      |  6
  0 |  0 | 0000 |  0
    |    |      |  6
    |    |      |  8
 10 | 10 | 0010 | 10
    |    |      |  2
    |    |      |  6
(48 rows)

-- anti join--
EXPLAIN (COSTS OFF) SELECT * FROM p1 t1 WHERE NOT EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;
                     QUERY PLAN                      
-----------------------------------------------------
 Nested Loop Anti Join
   Join Filter: (t1.a = t2.a)
   ->  Seq Scan on p1 t1
         Filter: (a <= 40)
   ->  Index Scan using p2_pkey on p2 t2
         Index Cond: (a = ANY (ARRAY[t1.a, $1, $2]))
(6 rows)

SELECT * FROM p1 t1 WHERE NOT EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;
 a  | b  |  c   
----+----+------
 38 | 13 | 0038
 16 | 16 | 0016
 40 | 15 | 0040
 32 |  7 | 0032
 10 | 10 | 0010
 34 |  9 | 0034
  4 |  4 | 0004
  2 |  2 | 0002
  8 |  8 | 0008
 14 | 14 | 0014
 22 | 22 | 0022
 26 |  1 | 0026
 28 |  3 | 0028
 20 | 20 | 0020
(14 rows)

-- semi join--
EXPLAIN (COSTS OFF) SELECT * FROM p1 t1 WHERE EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;
                     QUERY PLAN                      
-----------------------------------------------------
 Nested Loop Semi Join
   Join Filter: (t1.a = t2.a)
   ->  Seq Scan on p1 t1
         Filter: (a <= 40)
   ->  Index Scan using p2_pkey on p2 t2
         Index Cond: (a = ANY (ARRAY[t1.a, $1, $2]))
(6 rows)

SELECT * FROM p1 t1 WHERE EXISTS (SELECT 1 FROM p2 t2 WHERE t1.a = t2.a) AND t1.a <= 40;
 a  | b  |  c   
----+----+------
 12 | 12 | 0012
  6 |  6 | 0006
 36 | 11 | 0036
 18 | 18 | 0018
 30 |  5 | 0030
  0 |  0 | 0000
 24 | 24 | 0024
(7 rows)

set yb_nl_batch_size to 10;
set enable_seqscan to false;
explain (costs off) select * from p1 a join p2 b on a.a = b.a join p3 c on b.a = c.a join p4 d on a.b = d.b where a.b = 10 ORDER BY a.a, b.a, c.a, d.a;
                                                QUERY PLAN                                                 
-----------------------------------------------------------------------------------------------------------
 Sort
   Sort Key: a.a, d.a
   ->  Nested Loop
         ->  Index Scan using p4_pkey on p4 d
               Index Cond: (b = 10)
         ->  Materialize
               ->  Nested Loop
                     Join Filter: (a.a = b.a)
                     ->  Nested Loop
                           Join Filter: (a.a = c.a)
                           ->  Index Scan using p1_pkey on p1 a
                                 Index Cond: (b = 10)
                           ->  Index Scan using p3_a_idx on p3 c
                                 Index Cond: (a = ANY (ARRAY[a.a, $1, $2, $3, $4, $5, $6, $7, $8, $9]))
                     ->  Index Scan using p2_a_idx on p2 b
                           Index Cond: (a = ANY (ARRAY[a.a, $11, $12, $13, $14, $15, $16, $17, $18, $19]))
(16 rows)

select * from p1 a join p2 b on a.a = b.a join p3 c on b.a = c.a join p4 d on a.b = d.b where a.b = 10 ORDER BY a.a, b.a, c.a, d.a;
  a  | b  |  c   |  a  | b  |  c   |  a  | b  |  c   |  a  | b  |  c   
-----+----+------+-----+----+------+-----+----+------+-----+----+------
  60 | 10 | 0060 |  60 | 10 | 0060 |  60 | 10 | 0060 |  35 | 10 | 0035
  60 | 10 | 0060 |  60 | 10 | 0060 |  60 | 10 | 0060 | 210 | 10 | 0210
  60 | 10 | 0060 |  60 | 10 | 0060 |  60 | 10 | 0060 | 385 | 10 | 0385
  60 | 10 | 0060 |  60 | 10 | 0060 |  60 | 10 | 0060 | 560 | 10 | 0560
 210 | 10 | 0210 | 210 | 10 | 0210 | 210 | 10 | 0210 |  35 | 10 | 0035
 210 | 10 | 0210 | 210 | 10 | 0210 | 210 | 10 | 0210 | 210 | 10 | 0210
 210 | 10 | 0210 | 210 | 10 | 0210 | 210 | 10 | 0210 | 385 | 10 | 0385
 210 | 10 | 0210 | 210 | 10 | 0210 | 210 | 10 | 0210 | 560 | 10 | 0560
 360 | 10 | 0360 | 360 | 10 | 0360 | 360 | 10 | 0360 |  35 | 10 | 0035
 360 | 10 | 0360 | 360 | 10 | 0360 | 360 | 10 | 0360 | 210 | 10 | 0210
 360 | 10 | 0360 | 360 | 10 | 0360 | 360 | 10 | 0360 | 385 | 10 | 0385
 360 | 10 | 0360 | 360 | 10 | 0360 | 360 | 10 | 0360 | 560 | 10 | 0560
 510 | 10 | 0510 | 510 | 10 | 0510 | 510 | 10 | 0510 |  35 | 10 | 0035
 510 | 10 | 0510 | 510 | 10 | 0510 | 510 | 10 | 0510 | 210 | 10 | 0210
 510 | 10 | 0510 | 510 | 10 | 0510 | 510 | 10 | 0510 | 385 | 10 | 0385
 510 | 10 | 0510 | 510 | 10 | 0510 | 510 | 10 | 0510 | 560 | 10 | 0560
(16 rows)

DROP TABLE p1;
DROP TABLE p2;
DROP TABLE p3;
DROP TABLE p4;
